lib.rs

```diff
@@ -76,31 +47,76 @@ mod qjs {
             args: Vec<String>,
         ) -> Result<Output, String> {
             info!("evaluating js bytecode, code len: {}", bytecode.len());
-            eval(JsCode::Bytecode(&bytecode), args)
+            eval(&[Code::Bytecode(&bytecode)], args)
+        }
+
+        #[ink(message)]
+        pub fn eval_all(
+            &self,
+            codes: Vec<phat_js::Value>,
+            args: Vec<String>,
+        ) -> Result<Output, String> {
+            info!("batch evaluating {} scripts", codes.len());
+            let mut js_codes = Vec::new();
+            for code in &codes {
+                let js_code = match code {
+                    Value::String(s) => {
+                        info!("src len: {}", s.len());
+                        Code::Source(s)
+                    }
+                    Value::Bytes(b) => {
+                        info!("bytecode len: {}", b.len());
+                        Code::Bytecode(b)
+                    }
+                    Value::Undefined => return Err("undefined code".to_string()),
+                };
+                js_codes.push(js_code);
+            }
+            let output = eval(&js_codes, args)?;
+            Ok(output)
+        }
+
+        #[ink(message)]
+        pub fn compile(&self, js: String) -> Result<Vec<u8>, String> {
+            Ok(qjsbind::compile(&js, "<eval>")?)
         }
     }
 
-    fn eval(code: JsCode, args: Vec<String>) -> Result<Output, String> {
-        super::calc_and_set_code_hash(&code);
-        qjs_sys::eval(
-            &[
-                JsCode::Bytecode(BOOT_CODE),
-                JsCode::Source(&set_version()),
-                code,
-            ],
-            &args,
-        )
-        .map(Into::into)
+    fn eval(codes: &[Code], args: Vec<String>) -> Result<Output, String> {
+        unsafe { set_codes(codes) };
+
+        let rt = qjsbind::Runtime::new();
+        let ctx = rt.new_context();
+
+        setup_host_functions(&ctx)?;
+
+        let args = args.to_js_value(&ctx)?;
+        let global = ctx.get_global_object();
+        global.set_property("scriptArgs", &args)?;
+
+        ctx.eval(&Code::Bytecode(BOOT_CODE))?;
+        ctx.eval(&Code::Source(&set_version()))?;
+        let mut output = JsValue::undefined();
+        for code in codes.iter() {
+            output = ctx.eval(code)?;
+        }
+        if output.is_uint8_array() {
+            let bytes = output.decode_bytes()?;
+            return Ok(Output::Bytes(bytes));
+        }
+        if output.is_undefined() {
+            return Ok(Output::Undefined);
+        }
+        Ok(Output::String(output.to_string()))
     }
 
-    fn set_version() -> CString {
+    fn set_version() -> String {
         let version = env!("CARGO_PKG_VERSION");
-        CString::new(alloc::format!(
+        alloc::format!(
```

Comment:
Why don't use `this_crate::version_tuple!()` here?

Reply:
The code is older than this_crate exists.
Since the assigned `version` is a string, I think it is not necessary to switch to `this_crate` even for now.